### ISSUE TITLE
 412 'Precondition Failed'

### ISSUE THREAD

**Issue by RyanWor** [bachya/pytile#126]
**Describe the bug**
Tile integration within Home Assistant has not worked for me on two different HASS instances and the HA logs I did not find to be very helpful, so I downloaded pytile and attempted to run examples/test_api.py to test logging in with my username and password.

Upon running examples/test_api.py I get the following error:
`INFO:__main__:Error requesting data from tiles/X##############X: 412, message='Precondition Failed', url=URL('https://production.tile-api.com/api/v1/tiles/X##############X')`

Note: X##############X is (possibly?) a unique identifier which I redacted just in case, it is 26 characters and begins and ends with a letter with all numbers in between. 

**To Reproduce**
Followed the steps in this previously reported Tile issue in the home-assistant repo: 
https://github.com/home-assistant/core/issues#issuecomment-996165557

**Expected behavior**
Successful login to Tile account 

**Additional context**
I am not certain exactly when this stopped working, but it has been at least a few weeks now. 

Not sure if it is relevant, but the only thing that has changed on my Tile account in recent weeks as been adding a few new Tiles as well as some Tile Lost and Found Labels that they recently sent me for free. The Lost and Found Labels are a totally passive Tile product that allows someone who finds an item to contact the owner by scanning the QR code on the sticker. The shop up in the Tile app as another tracker but they lack many of the options of an active Tile tracker device. Perhaps these new Tile Lost and Found Labels are causing issues? 

https://www.tile.com/product/lost-and-found-labels

**Comment by RyanWor** [comment]
To add further troubleshooting and information around this issue. Today I tried a separate fresh TIle account with no devices or labels added and test_api.py woks and it successfully loaded the Tile integration within Home Assistant. I then added a new Lost and Found Label to this test account and both test_api.py fails and the Home Assistant integration fails when attempting to reload. Transferring the label to a different account then allows test_api.py and the Home Assistant integration to work again.

So I can say with near certainty that this is what is causing the issue. If it would be more appropriate for me to raise an issue on the Home Assistant GitHub tagging the Tile integration I can certainly do that, but my assumption was since pytile also seems affected by this issue that I should start here.

**Comment by bachya** [comment]
I, unfortunately, cannot reproduce this.

I spun up a fresh venv and ran [the example script](https://github.com/bachya/pytile/blob/dev/examples/test_api.py)—everything looked as it should:

```
INFO:__main__:Tile Count: 7
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2017-12-07 02:53:35.149000
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2022-06-22 03:56:48.900000
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2021-10-07 19:32:41.930000
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2020-02-17 05:45:56.716000
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2022-10-13 22:16:12.014000
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2019-12-25 05:05:23
INFO:__main__:UUID: REDACTED
INFO:__main__:Name: REDACTED
INFO:__main__:Type: TILE
INFO:__main__:Latitude: REDACTED
INFO:__main__:Longitude: REDACTED
INFO:__main__:Last Timestamp: 2020-08-14 04:55:18.657000
```

It sounds like these labels are the root of your issue, but I'm not clear what they refer to—are these labels that get assigned in the Tile mobile app? If so, could you show some screenshots of that workflow?

**Comment by RyanWor** [comment]
Yes they show up like another Tile in the Tile app, but they lack many of the functions of a active Tile BLE type device. I can take a screen recording of what setting up one in the Tile app looks like if that would be helpful? Or just a screen recording of the options of one of the labels in the Tile app. Alternatively, I could provide credentials to a spare Tile test account with a label already added, or even transfer a label to an account you already control, if either of those would be helpful. I've attached a verbose log when running test_api.py, I have REDACTED the Tile's UUID in this copy but can send an unredacted version privately if that would be helpful. Thanks for your help!

[ryanw@Desktop-Z490 ~pytile.txt](https://github.com/bachya/pytile/files/10437006/ryanw%40Desktop-Z490.pytile.txt)

**Comment by bachya** [comment]
Thanks for the detail! Could you run [`test_api.py`](https://github.com/bachya/pytile/blob/dev/examples/test_api.py) again, but this time, with line 18 changed:

```python
logging.basicConfig(level=logging.DEBUG)
```

Totally fine to scrub UUIDs/other sensitive info before posting back here.

**Comment by RyanWor** [comment]
Sure thing:

`DEBUG:pytile:Data received from /clients/REDACTED: {'version': 1, 'revision': 1, 'timestamp': '2023-01-17T23:57:21.788Z', 'timestamp_ms': 1673999841788, 'result_code': 0, 'result': {'locale': 'en-US', 'entityName': 'CLIENT', 'client_uuid': 'REDACTED', 'app_id': 'ios-tile-production', 'app_version': '2.89.1.4774', 'os_name': None, 'os_release': None, 'model': None, 'signed_in_user_uuid': None, 'registration_timestamp': 1673999841774, 'user_device_name': None, 'beta_option': False, 'last_modified_timestamp': 1673999841774}}
DEBUG:pytile:Data received from /clients/REDACTED/sessions: {'version': 1, 'revision': 1, 'timestamp': '2023-01-17T23:57:22.131Z', 'timestamp_ms': 1673999842131, 'result_code': 0, 'result': {'client_uuid': 'REDACTED', 'user': {'universalContactEmail': 'test@email.com', 'universalContactPhone': '+1 555-555-5555', 'universalContactPhoneRegionCode': 'US', 'user_uuid': 'REDACTED', 'full_name': None, 'email': 'REDACTED', 'beta_eligibility': False, 'gift_recipient': False, 'locale': 'en-US', 'email_shared': True, 'image_url': None, 'email_change_request': None, 'linked_accounts': [], 'shipping_address_timestamp': None, 'pw_exists': True, 'status': 'ACTIVATED', 'registration_timestamp': 1673971851426, 'last_modified_timestamp': 1673999837433}, 'session_start_timestamp': 1673999842114, 'session_expiration_timestamp': 1674021442114, 'changes': 'EXISTING_ACCOUNT'}}
DEBUG:pytile:Data received from /tiles/tile_states: {'version': 1, 'revision': 1, 'timestamp': '2023-01-17T23:57:22.185Z', 'timestamp_ms': 1673999842185, 'result_code': 0, 'result': [{'location': {'latitude': REDACTED, 'longitude': REDACTED, 'location_timestamp': 1673834466356, 'horizontal_accuracy': 25.0, 'client_name': None}, 'tile_id': 'REDACTED', 'mark_as_lost': {'timestamp': -1, 'is_lost': False, 'is_owner_contact_provided': True, 'owner_contact_message': 'This item has been lost, please contact me. Thank you!'}},{'location': {'latitude': REDACTED, 'longitude': REDACTED, 'location_timestamp': 1673999789661, 'horizontal_accuracy': 25.0, 'client_name': None}, 'tile_id': 'REDACTED', 'mark_as_lost': {'timestamp': -1, 'is_lost': False, 'is_owner_contact_provided': False, 'owner_contact_message': None}}]}
INFO:__main__:Error requesting data from tiles/REDACTED: 412, message='Precondition Failed', url=URL('https://production.tile-api.com/api/v1/tiles/REDACTED')`

**Comment by bachya** [comment]
I can't see from your screenshots how exactly you assign a label to a Tile—could you clarify?

**Comment by RyanWor** [comment]
You do not assign a label to a Tile. It is it's own item in your Tile "inventory". Newer Tiles feature similar QR codes with the same functionality as the labels (ability to scan QR code and email/text the owner) but the labels only have the ability to add the contact info; you cannot actually see their location or anything like a real Tile since they are purely a sticker with a QR code.

**Comment by RyanWor** [comment]
Would it be helpful for me to send you a fresh label QR code for you to see how they work and troubleshoot with?

**Comment by bachya** [comment]
I'm open to your suggestion. It sounds like you're experiencing an issue when these labels are introduced, so anything that can assist me in testing that use case will help here.

**Comment by RyanWor** [comment]
Yeah not a problem. I think being able to see how they work will probably help. Do you want to shoot me an email at ryanworrell@gmail.com and I'll reply with a picture of one I've reset so you can scan it with the Tile app and add it?

**Comment by bachya** [comment]
Just emailed you.

**Comment by RyanWor** [comment]
Just sent you a QR code back. Thanks again!

**Comment by RyanWor** [comment]
Bumping thread to prevent auto-closure. Issue is still outstanding and pending resolution.

**Comment by RyanWor** [comment]
Hi @bachya do you want me to keep this issue open? Not sure if you have had a chance to look at this yet.

**Comment by bachya** [comment]
Yes, please keep it open. I haven't had the bandwidth to look into it, but will as soon as I'm able.

**Comment by bit-bat** [comment]
I can confirm that the tile Lost and Found QR labels are causing an issue. I was seeing constant error with home assistant integration, and as soon as I removed the Lost and Found label from my tile account (thanks to the thread for the idea), integration was successful.

**Comment by RyanWor** [comment]
Thanks @bit-bat 

Fwiw, I recommend the temporary workaround of keeping your labels in a separate Tile account. Tile accounts are free anyways and beyond setting the owner information there is really nothing more that needs to be done in the app with a Lost & Found label once it is setup.

**Comment by bachya** [comment]
I finally had some time to look into this—thanks for your patience (and assistance with getting access to one of these labels), @RyanWor.

The `/api/v1/tiles/tile_states` endpoint returns this label alongside all of the other "real" tiles—it's response payload looks like this:

```json
{
  "location": {
    "latitude": -90.1,
    "longitude": -180.1,
    "location_timestamp": 1681151307255,
    "horizontal_accuracy": 25,
    "client_name": null
  },
  "tile_id": "REDACTED",
  "mark_as_lost": {
    "timestamp": -1,
    "is_lost": false,
    "is_owner_contact_provided": true,
    "owner_contact_message": "This item has been lost, please contact me. Thank you!"
  }
},
```

Calling `/api/v1/tiles/<TILE_UUID>` results in the `HTTP 412` that you see—in a way, I'd expect this since that endpoint exists to provide additional Tile details, and in the case of a label, there aren't any additional details. I'm guessing that the failed "precondition" is us calling this endpoint with an invalid Tile type.

I could modify our use of `api/v1/tiles/<TILE_UUID>` to ignore label-Tiles if I could confirm how to identify one. In looking at the response payload, the only differences between the label and the other Tiles are:

1. The `is_owner_contact_provided` field is `true`.
2. The `owner_contact_message` field has a value.

So, can we safely assume that non-label Tiles _can't_ have these values? I dug around the app and couldn't find a way to apply an `owner_contact_message` to a non-label Tile, but that doesn't mean it's impossible...

**Comment by bachya** [comment]
After digging further, the Tile Labels don't conform to many of the same structural standards we expect from other Tiles. So, one way to solve this is to ignore Tile Labels altogether. That seems safe, given that Labels can't ever update their state—is there any reason not to go that route?

**Comment by RyanWor** [comment]
Actually, newer gen Tiles that feature the "SCAN IF FOUND" QR code on them are capable of leveraging both of those fields. Although they are only used when you set the Tile to "Notify When Found" in the Tile app and provide contact information. 

I just tested with a recent gen Tile Pro that has the "SCAN IF FOUND" QR code printed on the case and confirmed both is_owner_contact_provided changed to true and owner_contact_message was populated with the message set in the Tile app when marking the Tile "Notify When Found". 

Now while this is a specific use case may not occur very often, I think most would prefer that if an item was lost and "Notify When Found" was turned on, on the chance that it did get picked up again, I would prefer pytile to still report it's location. One of the best features of pytile and the integration with Home Assistant is being able to capture the location history (30 day location history is available in the Tile app but only with a Premium subscription).

Just throwing out some possible ideas for alternatives: 

1) If the Tile stickers lack many of the other expected values of a real tile (like maybe longitude and latitude) could we filter them out based on the lack of those value fields even existing? 

2) Do we know enough about the formatting of tile_id to be able to differentiate between true Tile tracker devices and Tile stickers? 

Thanks again for your help on this @bachya, much appreciated!

**Comment by bachya** [comment]
Thanks for your thoughts, @RyanWor. Correct, my new thinking would be to loop through each Tile, perform our usual API requests, and if we get an `HTTP 412` at any point, we ignore that Tile.

It's still possible I could figure out a way to incorporate Labels, but the official API is really messy in that regard; if ignoring them is functionally feasible, it would certainly be easier on me.

**Comment by RyanWor** [comment]
Happy to help @bachya 

For what it's worth, I don't see any benefit from my perspective on having the Labels even show up. That said, my sole use for pytile to date is tracking location in Home Assistant. So there could be others that find some reason to want to see Labels in pytile, but personally I can't see any good use for them in there at this time beyond just having an inventory of all Tiles including Labels registered to one's account.

**Comment by bachya** [comment]
Got it. I'll roll forward with ignoring them, and if another use case comes along at some point, I can revisit. Thanks!

**Comment by bachya** [comment]
@RyanWor FYI, I've submitted a HASS PR that includes this fix: https://github.com/home-assistant/core/pull/91191

**Comment by RyanWor** [comment]
> @RyanWor FYI, I've submitted a HASS PR that includes this fix: [home-assistant/core#91191](https://github.com/home-assistant/core/pull/91191)

Awesome, looks like it was merged into 2023.4.3 update. Thanks again for your help on this!
